torch/_subclasses/fake_tensor.py

```diff
@@ -965,6 +977,163 @@ def should_allow_numbers_as_tensors(func: OpOverload):
     )
 
 
+_DISPATCH_CACHE = {}
+
+
+class _Boolean(NamedTuple):
+    value: bool
+
+
+_True = _Boolean(True)
+_False = _Boolean(False)
+
+
+class FakeTensorMetadata(NamedTuple):
+    size: Tuple[int, ...]
+    stride: Tuple[int, ...]
+    dtype: torch.dtype
+    device: torch.device
+    requires_grad: bool
+    is_view: bool
+    is_inference: bool
+    constant: Any
+
+
+def _dispatch_should_fallback(op: Any) -> bool:
+    if op == torch.ops.prim.device.default:
+        return True
+
+    assert isinstance(op, OpOverload)
+    schema = op._schema
+
+    # Return true if there's any alias information in any of the
+    # return values.
+    return len(schema.returns) > 0 and any(
+        r.alias_info is not None for r in schema.returns
+    )
+
+
+def _is_notsupported_metadata(thing: Any) -> bool:
+    return isinstance(thing, (SymTypes, torch.ScriptObject)) or (isinstance(thing, torch.Tensor) and thing.is_sparse)
+
+
+def _apply(f: Callable, iterable) -> Any:
+    elems = (f(el) for el in iterable)
+    # Expect tuple_or_list to be either:
+    if isinstance(iterable, list) or type(iterable) is tuple:
+        # - a list
+        # - a tuple
+        return type(iterable)(elems)
+    elif hasattr(iterable, "_fields"):
+        # - a namedtuple
+        return type(iterable)(*elems)
+    raise ValueError(f"unsupported tuple/list-like type: {type(iterable)}")
+
+
+def _to_metadata(output: Any, allow_views: bool = False) -> Any:
+    if isinstance(output, (torch.Tensor, FakeTensor)):
+        assert allow_views or not output._is_view(), "can't handle tensor views"
+        return FakeTensorMetadata(
+            size=output.size(),
+            stride=output.stride(),
+            dtype=output.dtype,
+            device=output.device,
+            is_view=output._is_view(),
+            is_inference=output.is_inference(),
+            requires_grad=output.requires_grad,
+            constant=getattr(output, "constant", None),
+        )
+    elif isinstance(output, bool):
+        return {True: _True, False: _False}[output]
+    elif isinstance(
+        output,
+        (
+            bool,
+            int,
+            float,
+            complex,
+            str,
+            type(None),
+            torch.dtype,
+            torch.device,
+            torch.layout,
+            torch.memory_format,
+        ),
+    ):
+        return output
+    elif isinstance(output, (tuple, list)):
+        return _apply(partial(_to_metadata, allow_views=allow_views), output)
+
+    raise ValueError(f"unsupported type to metadata: {type(output)}")
+
+
+def _from_metadata(metadata: Any, fake_mode) -> Any:
+    if isinstance(metadata, FakeTensorMetadata):
+        with torch.inference_mode(mode=metadata.is_inference):
```

Comment:
I guess I have a similar question about whether we need to preserve inference_mode here.